src/System.Management.Automation/engine/Subsystem/SubsystemManager.cs

```diff
@@ -0,0 +1,254 @@
+        /// </summary>
+        /// <param name="kind">A specific <see cref="SubsystemKind"/>.</param>
+        /// <returns>The <see cref="SubsystemInfo"/> object that represents the concrete subsystem.</returns>
+        public static SubsystemInfo GetSubsystemInfo(SubsystemKind kind)
+        {
+            if (s_subSystemKindMap.TryGetValue(kind, out SubsystemInfo subsystemInfo))
+            {
+                return subsystemInfo;
+            }
+
+            throw new ArgumentException(
+                StringUtil.Format(
+                    SubsystemStrings.SubsystemKindUnknown,
+                    kind.ToString()));
+        }
+
+        #endregion
+
+        #region public - Subsystem registration
+
+        /// <summary>
+        /// Subsystem registration.
+        /// </summary>
+        /// <typeparam name="TConcreteSubsystem">The concrete subsystem base type.</typeparam>
+        /// <typeparam name="TImplementation">The implementation type of that concrete subsystem.</typeparam>
+        /// <param name="proxy">An instance of the implementation.</param>
+        public static void RegisterSubsystem<TConcreteSubsystem, TImplementation>(TImplementation proxy)
+            where TConcreteSubsystem : class, ISubsystem
+            where TImplementation : class, TConcreteSubsystem
+        {
+            if (proxy == null)
+            {
+                throw new ArgumentNullException(nameof(proxy));
+            }
+
+            RegisterSubsystem(GetSubsystemInfo(typeof(TConcreteSubsystem)), proxy);
+        }
+
+        /// <summary>
+        /// Register an implementation for a subsystem.
+        /// </summary>
+        /// <param name="kind">The target <see cref="SubsystemKind"/> of the registration.</param>
+        /// <param name="proxy">An instance of the implementation.</param>
+        public static void RegisterSubsystem(SubsystemKind kind, ISubsystem proxy)
+        {
+            if (proxy == null)
+            {
+                throw new ArgumentNullException(nameof(proxy));
+            }
+
+            if (kind != proxy.Kind)
+            {
+                throw new ArgumentException(
+                    StringUtil.Format(
+                        SubsystemStrings.ImplementationMismatch,
+                        proxy.Kind.ToString(),
+                        kind.ToString()),
+                    nameof(proxy));
+            }
+
+            RegisterSubsystem(GetSubsystemInfo(kind), proxy);
+        }
+
+        private static void RegisterSubsystem(SubsystemInfo subsystemInfo, ISubsystem proxy)
+        {
+            if (subsystemInfo.RequiredCmdlets.Any() || subsystemInfo.RequiredFunctions.Any())
+            {
+                // Process 'proxy.CmdletImplementationAssembly' and 'proxy.FunctionsToDefine'
+                // Functions are added to global scope.
+                // Cmdlets are loaded in a way like a snapin, making the 'Source' of the cmdlets to be 'Microsoft.PowerShell.Core'.
```

Comment:
Can you expand on this?  It is not clear how registered subsystem cmdlets are loaded into Core namespace.

Reply:
I haven't got to the implementation details yet, but it should be doable.